Update Workload API spec with JWT-SVID Profile

azdagron opened the pull request:
This PR adds the JWT-SVID Profile to the Workload API specification.

Most of the JWT-SVID text came from a draft proposal created by @evan2645. There have been some small tweaks, mostly adapting the language based on the recent addition of `FetchX509Bundles`, which closely mirrors the `FetchJWTBundles` RPC.

I also took the opportunity to create a canonical, complete definition of the Workload API gRPC Service definition. Consequently some of the service definition text in the spec has been adjusted.

Commit:
Update Workload API spec with JWT-SVID Profile

This PR adds the JWT-SVID Profile to the Workload API specification.

Most of the JWT-SVID text came from a draft proposal created by
@evan2645. There have been some small tweaks, mostly adapting the
language based on the recent addition of FetchX509Bundles, which closely
mirrors the FetchJWTBundles RPC.

I also took the opportunity to create a canonical, complete definition
of the Workload API gRPC Service definition. Consequently some of the
service definition text in the spec has been adjusted.

Signed-off-by: Andrew Harding <aharding@vmware.com>

azdagron commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,51 +10,43 @@ Portable and interoperable cryptographic identity for networked workloads is per
 The SPIFFE Workload API is an API which provides information and services that enable workloads, or compute processes, to leverage SPIFFE identities and SPIFFE-based authentication systems. It is served by the [SPIFFE Workload Endpoint](SPIFFE_Workload_Endpoint.md), and comprises a number of services, or *profiles*.
 
-Currently, only one profile is defined: the [X.509-SVID Profile](#5-x509-svid-profile). As such, supporting this profile is mandatory. Future versions of this specification may introduce additional profiles, and the X.509-SVID Profile may become optional.
+Currently, there are two profiles, both of which are mandatory:
```
> both of which are mandatory

I couldn't remember where the group landed in terms of optionality. Happy to amend this if it isn't accurate.

azdagron replied:
@evan2645 reminded me of what we discussed in SIG-SPEC a bit ago, that the profile is mandatory but can be administratively disabled.

spikecurtis commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -169,6 +182,134 @@ If the server redacts all SVIDs from a workload, it SHOULD send the "PermissionD
+
+The FetchJWTBundles method is different. It is implemented as a gRPC server-side stream in order to facilitate rapid propagation of updates, namely key introductions or redactions. Every response message sent by the server MUST include the full set of information, and not just the information which has changed. This avoids complexity associated with state tracking on both Client and Server implem ...
+
+The client and server behavior of FetchJWTBundles is identical to that of the X509-SVID profile. Please see the Client and Server Behavior section of the X509-SVID profile for more detailed information.
```
I find this sentence a little hard to understand because you're comparing an RPC to a "profile" so it feels like apples and oranges to say they are identical.  The term behavior is also a little vague: clearly the RPCs needs to do some things that are different from one another, otherwise they would all be the same RPC, so I think we need to get a little more precise about what exactly is the same about them.

azdagron replied:
Yes, I agree this is really unwieldy. In the larger refactoring I had started, there was only one section that described the streaming behavior, since what is mentioned in the X.509 profile is really just a property of all of the streaming RPCs in the Workload API. I'll see if I can't word smith this a little.

spikecurtis commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -169,6 +182,134 @@ If the server redacts all SVIDs from a workload, it SHOULD send the "PermissionD
+
+The JWTBundlesResponse message includes a map of JWKS bundles, keyed by trust domain. When validating a JWT-SVID, the validator should use the bundle corresponding to the trust domain of the subject. If a JWT bundle for the specified trust domain is not present, then the token is untrusted.
+
+Please note that nominally, workloads will use the ValidateJWTSVID method for JWT validation, allowing the SPIFFE Workload API to perform validation on their behalf. Doing this removes the need for the workload to implement validation logic, which can be error prone.
```
```suggestion
Workload API clients SHOULD use the ValidateJWTSVID method for JWT validation if supported by the client, allowing the SPIFFE Workload API to perform validation on their behalf. Doing this removes the need for the workload to implement validation logic, which can be error prone.
```

azdagron replied:
:+1:

azdagron commented:
Something came up as I was reviewing SPIRE's implementation of `FetchJWTSVID` that I'd love clarification on. If a workload is authorized for some set of identities, but the workload asks for a JWT-SVID with a specific SPIFFE ID, that of which the workload is not authorized for, what should the behavior of the RPC be?

SPIRE currently returns an empty response in this instance. This provides an unambiguous response. If however, the RPC returned PERMISSION_DENIED, it would be difficult for the workload to understand whether it wasn't authorized for any identity, or just the specific identity being requested.

This text should clarify the RPC behavior. Curious what thoughts others have.

spikecurtis commented:
APIs that I have worked with use PERMISSION_DENIED to indicate that the specific request is denied; not to indicate that all requests of that type are denied.  So, I don't think it would be confusing if a workload asks for a specific SPIFFE ID and gets denied.

We specifically have text that says that `JWTSVIDResponse` is not allowed to be empty.

Commit:
Further reorganization

- Moves common phrasing to a profile-agnostic Client and Server behavior
section.
- Introduces a consistent framing for the profile definition and RPC
details.

Signed-off-by: Andrew Harding <aharding@vmware.com>

azdagron commented:
Sorry it took so long to address the comments. It became apparent, to me at least, that it was going to be difficult to fix the phrasing in the current PR without the reorganization of sections that I alluded to several months ago on the SIG-SPEC call. I decided to bite the bullet and proceed with a slightly aggressive reorganization. I've moved all of the profile-agnostic materials into a common "Client and Server Behavior" section. I've also cleaned up and provided a uniform structure for defining the profile and its associated RPCs.

The change is kind of big, and for that I'm sorry. I couldn't see a clearer way to proceed though. Any feedback is welcome, up to and including "no way, start over" :)

spikecurtis commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+
+All fields in the `JWTSVID` message are mandatory.
+
+If the client is not authorized for any identities, nor the specific identity requested via the `spiffe_id` field, then the server SHOULD respond with the "PermissionDenied" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information).
```
```suggestion
If the client is not authorized for any identities, or not authorized for the specific identity requested via the `spiffe_id` field, then the server SHOULD respond with the "PermissionDenied" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information).
```

azdagron replied:
:+1:

spikecurtis commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+#### 5.2.1 FetchX509SVID
 
-The only mandatory field in the `X509SVIDResponse` message is the `svids` field. If the client is not entitled to an SVID, then the server SHOULD respond with the "PermissionDenied" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information). The `crl` field and the `federated_bundles` field are opt ...
+The `FetchX509SVID` RPC streams back X509-SVIDs, and X.509 bundles for authenticated X509-SVIDs for both the trust domain in which the server resides and foreign trust domains.
```
```suggestion
The `FetchX509SVID` RPC streams back X509-SVIDs, and X.509 bundles for both the trust domain in which the server resides and foreign trust domains. These bundles MUST only be used to authenticate X.509-SVIDs.
```

azdagron replied:
:+1:

spikecurtis commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+
+#### 6.2.2 FetchJWTBundles
+
+The `FetchJWTBundles` RPC streams back JWT bundles for authenticated JWT-SVIDs for both the trust domain in which the server resides and foreign trust domains.
```
```suggestion
The `FetchJWTBundles` RPC streams back JWT bundles for both the trust domain in which the server resides and foreign trust domains. These bundles MUST only be used to authenticate JWT-SVIDs.
```

azdagron replied:
:+1:

spikecurtis reviewed (dismissed):
Couple minor suggested changes for clarity.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+- [JWT-SVID Profile](#6-jwt-svid-profile)
 
-The SPIFFE Workload API MUST NOT be extended beyond this specification. Implementers wishing to provide extended functionality may do so by introducing new gRPC services, according to the [extensibility method](SPIFFE_Workload_Endpoint.md#7-extensibility-and-services-rendered) outlined in the SPIFFE Workload Endpoint specification.
+Both profiles are mandatory and MUST be implemented. However, one or the other MAY be administratively disabled.
```
I'm not sure how to reconcile these two sentences. If a profile may be disabled, then why _must_ the profile be implemented?

spikecurtis replied:
I think this is down to lack of specificity _on whom and where_ we are imposing specification requirements, i.e. _implementers_ are required to implement both to claim their _implementation_ meets the spec.  _Operators_ are not required to enable both on any specific _deployment_.

azdagron replied:
I added some clarifying language.

justinburke commented on standards/SPIFFE_Workload_API.md:
````diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
 
-```protobuf
-syntax = "proto3";
+The SPIFFE Workload API MUST NOT be extended beyond this specification. Implementers wishing to provide extended functionality may do so by introducing new gRPC services, according to the [extensibility method](SPIFFE_Workload_Endpoint.md#7-extensibility-and-services-rendered) outlined in the SPIFFE Workload Endpoint specification.
````
Out of curiosity, why must the API not be extended? Is is to preserve cross-implementation compatibility? (Might want to explain the reasoning here in the text.)

azdagron replied:
Good catch! This is text is unrelated to the JWT-SVID profile addition (or subsequent rejiggering of the text). Can I suggest we defer edits here to a subsequent PR?

azdagron replied:
Talked IRL; punting to another PR.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+### 4.4 Default Values and Redacted Information
 
-The `FetchX509SVID` RPC will always provide a Trust Bundle for the Trust Domain in which an SVID resides, however, it may also provide bundles for foreign Trust Domains.
+SPIFFE Workload API clients may at times encounter fields in the response message that have a default value, or may notice that information included in a previous response is not included in the latest response. For instance, a client may encounter a default value in the `federated_bundles` field after having previously received a federated bundle.
```
Consider rephrasing the first sentence to something like:
SPIFFE Workload API response messages are complete updates to previously sent response messages. When a response message contains fields which are set to default or empty values, clients MUST interpret the values of those fields to have been set to their default or empty values; previously received, non-default or non-empty values MUST NOT be retained by a client after receiving a default or empty value for the fields.

azdagron replied:
Awesome. I took this verbatim and also tweaked following sentence to fill out the example a little more.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+Various RPCs defined in this specification can return trust bundles for foreign trust domains. Inclusion of foreign bundles enables workloads to communicate *across* trust domains, and is the primary mechanism through which federation is enabled. A bundle representing a foreign trust domain is known as a *Federated Bundle*.
 
-In order to support the widest variety of use cases, the X.509-SVID Profile supports the issuance of multiple identities, while also defining a default identity. It is expected that workloads which are aware of multiple identities can handle decision making on their own. Workloads which don’t understand how to leverage multiple identities may use the default identity. The default identity is the  ...
+When authenticating a client from a foreign trust domain, the authenticator chooses the bundle representing the client’s presented trust domain for validation. Similarly, when authenticating a server, the client uses the bundle representing the server’s trust domain. If no matching bundle is present for the SVID in use, then the peer is untrusted. This approach is required in order to account for ...
```
Would the first sentence be just as true if written:
When authenticating a client, the authenticator ...

What caught my attention is that the next sentence does not contain the phrase "from a foreign trust domain".

azdagron replied:
I think so. Fixed.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+Various RPCs defined in this specification can return trust bundles for foreign trust domains. Inclusion of foreign bundles enables workloads to communicate *across* trust domains, and is the primary mechanism through which federation is enabled. A bundle representing a foreign trust domain is known as a *Federated Bundle*.
 
-In order to support the widest variety of use cases, the X.509-SVID Profile supports the issuance of multiple identities, while also defining a default identity. It is expected that workloads which are aware of multiple identities can handle decision making on their own. Workloads which don’t understand how to leverage multiple identities may use the default identity. The default identity is the  ...
+When authenticating a client from a foreign trust domain, the authenticator chooses the bundle representing the client’s presented trust domain for validation. Similarly, when authenticating a server, the client uses the bundle representing the server’s trust domain. If no matching bundle is present for the SVID in use, then the peer is untrusted. This approach is required in order to account for ...
```
I recommend either explaining the relevance of support for SAN URI Name Constraints or remove reference it from this paragraph.

My preference would be to remove the reference since I think it's a distraction here, even if the relevance was explained.

azdagron replied:
I'm OK removing the reference. It is part of the original text however. @spikecurtis, @evan2645 are you also comfortable removing this reference?

azdagron replied:
Talked IRL. Punting this to a separate PR.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+
+    // Fetch X.509-SVIDs for all SPIFFE identities the workload is entitled to,
+    // as well as related information like trust bundles and CRLs. As this
+    // information changes, subsequent messages will be sent.
```
Instead of "sent", consider "streamed from the server".

azdagron replied:
Updated all instances.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+
+    // Fetch trust bundles and CRLs. Useful for clients that only need to
+    // validate SVIDs without obtaining an SVID for themself. As this
+    // information changes, subsequent messages will be sent.
```
Same comment as above, consider s/sent/streamed from the server/.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+
+
+// The X509SVIDRequest message conveys parameters for requesting an X.509-SVID.
+// There are currently no such parameters.
```
s/such/request/

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+message X509SVIDRequest {  }
+
+// The X509SVIDResponse message carries X.509-SVIDs and related information,
+// including a global CRL and list of bundles the workload is federated with.
```
Consider replacing "list of bundles the workload is federated with" with
"a list of foreign trust bundles which the workload may use to federate with".

1) s/list/a list/
2) are the bundles necessarily foreign bundles? I'm not sure, please clarify.
3) does the presence of the bundles imply that the workload "is federated with" them? or merely helps enable that federation capability?

azdagron replied:
I took a stab at rewording this.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
 
-    // ASN.1 DER encoded certificate chain. MAY include intermediates,
-    // the leaf certificate (or SVID itself) MUST come first.
+    // Required. ASN.1 DER encoded certificate chain. MAY include
```
While it's clear that the leaf cert must be at the head of the list, does the order of the intermediates (following the leaf) matter?

azdagron replied:
Not in my experience, no.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
 }
+
+message JWTSVIDRequest {
+    // Required. The audience the workload intends to authenticate against.
```
Since the field is marked `repeated`, should "audience" be "audience(s)" ?

azdagron replied:
:+1:

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,351 @@ Portable and interoperable cryptographic identity for networked workloads is per
+    repeated string audience = 1;
+
+    // Optional. The requested SPIFFE ID for the JWT-SVID. If unset, JWT-SVIDs
+    // for all identities the workload is entitled to are returned.
```
Since this comment belongs to the JWTSVIDRequest (not the Response), I think we should describe the semantics of the request as opposed to expected values in the response. Consider rephrasing the second sentence to something like: If unset, all JWT-SVIDs to which the workload is entitled are requested.

azdagron replied:
I like it.

justinburke reviewed (changes requested):
I got about half way through my review in this session. Posting my comments so far.

hug-dev commented:
Hello 👋!

Just showing our interest, in Parsec, to see the JWT SVID profile merged in the Workload API. This is to give us a bit more confidence that the JWT-SVID API are standardised and stable. We are currently consuming it via [`rust-spiffe`](https://github.com/heavypackets/rust-spiffe/blob/master/spiffe/src/workload/workload_api.proto) and using the same API as in [`go-spiffe`](https://github.com/spiffe/go-spiffe/blob/main/v2/proto/spiffe/workload/workload.proto). Comments on this PR do not seem to be on the API itself so I think we are fine.

~~Also, showing our interest in having the complete workload API, committed in this repository as an independant `.proto` file. That way different SPIFFE clients can directly fetch it from the "official" source (at specific versions) in their build script. Currently `rust-spiffe` is copying it but that's not ideal I guess.~~ That's exactly what you did 😅 Awesome!

Commit:
Address feedback

Signed-off-by: Andrew Harding <aharding@vmware.com>

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,354 @@ Portable and interoperable cryptographic identity for networked workloads is per
 }
 
-// The X509BundlesResponse message carries a set of global CRLs and a
+// The X509BundlesResponse message carries a global CRL and a
```
The `crl` field is still marked as repeated but the description here has dropped the notion of "a set of". This seems inconsistent.

azdagron replied:
Good catch. I'll restore the original language and update the field descriptions for consistency.

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,354 @@ Portable and interoperable cryptographic identity for networked workloads is per
+Workload API clients SHOULD use the `ValidateJWTSVID` method for JWT validation if supported by the client, allowing the SPIFFE Workload API to perform validation on their behalf. Doing this removes the need for the workload to implement validation logic, which can be error prone.
 
-If the server redacts all SVIDs from a workload, it SHOULD send the "PermissionDenied" gRPC status code (terminating the gRPC response stream). The client SHOULD cease using the redacted SVIDS. The client MAY attempt to reconnect with another call to the `FetchX509SVID` RPC after a backoff.
+When interfacing with legacy JWT validators, the `FetchJWTBundles` method can be used to fetch JWKS bundles that can be used to validate JWT-SVID signatures. For instance, if the SPIFFE Workload API is available but the JWT validating software is not aware of the Workload API, it is possible to write a small shim that can retrieve the bundles and feed them to the legacy workload.
```
```suggestion
When interfacing with legacy JWT validators, the `FetchJWTBundles` method can be used to fetch JWKS bundles that can be used to validate JWT-SVID signatures. For instance, if the SPIFFE Workload API is available but the JWT validating software is not aware of the Workload API (and thus cannot call `ValidateJWTSVID`), implementations can instead individually retrieve each bundle and feed them to the legacy workload for validation.
```

azdagron replied:
:+1:

justinburke commented on standards/SPIFFE_Workload_API.md:
```diff
@@ -10,164 +10,354 @@ Portable and interoperable cryptographic identity for networked workloads is per
+When interfacing with legacy JWT validators, the `FetchJWTBundles` method can be used to fetch JWKS bundles that can be used to validate JWT-SVID signatures. For instance, if the SPIFFE Workload API is available but the JWT validating software is not aware of the Workload API, it is possible to write a small shim that can retrieve the bundles and feed them to the legacy workload.
 
-If the server redacts all trust bundles from a client using the `FetchX509Bundles` RPC, it SHOULD send the "PermissionDenied" gRPC status code (terminating the gRPC response stream). The client SHOULD cease using the redacted trust bundles. The client MAY attempt to reconnect with another call to the `FetchX509Bundles` RPC after a backoff.
+The `FetchJWTBundles` method returns bundles keyed by the SPIFFE ID of the trust domain. When validating a JWT-SVID, the validator should use the bundle corresponding to the trust domain of the subject. If a JWT bundle for the specified trust domain is not present, then the token is untrusted.
```
```suggestion
The `FetchJWTBundles` method returns bundles keyed by the SPIFFE ID of the trust domain. When validating a JWT-SVID, the validator MUST use the bundle corresponding to the trust domain of the subject. If a JWT bundle for the specified trust domain is not present, then the token is untrusted.
```

azdagron replied:
Nice catch.

Commit:
Address more comments

Signed-off-by: Andrew Harding <aharding@vmware.com>

Commit:
Merge branch 'main' into add-jwt-svid-profile

hug-dev commented:
Nice! Thank you.
